tools/libSwiftScan/SwiftCaching.cpp

```diff
@@ -0,0 +1,902 @@
+            .str()
+            .c_str());
+
+  auto *Result = std::get_if<clang::cas::CompileJobCacheResult>(&Comp->Result);
+  assert(Result && "unexpected variant");
+  switch (Result->getOutput(idx).Kind) {
+  case clang::cas::CompileJobCacheResult::OutputKind::MainOutput:
+    return swift::c_string_utils::create_clone(
+        swift::file_types::getTypeName(swift::file_types::TY_ClangModuleFile)
+            .str()
+            .c_str());
+  case clang::cas::CompileJobCacheResult::OutputKind::Dependencies:
+    return swift::c_string_utils::create_clone(
+        swift::file_types::getTypeName(swift::file_types::TY_Dependencies)
+            .str()
+            .c_str());
+  case clang::cas::CompileJobCacheResult::OutputKind::SerializedDiagnostics:
+    return swift::c_string_utils::create_clone(
+        swift::file_types::getTypeName(swift::file_types::TY_CachedDiagnostics)
+            .str()
+            .c_str());
+  }
+}
+
+swiftscan_cached_output_t
+swiftscan_cached_compilation_get_output(swiftscan_cached_compilation_t id,
+                                        unsigned idx) {
+  auto *Comp = unwrap(id);
+  if (auto *Result =
+          std::get_if<swift::cas::CompileJobCacheResult>(&Comp->Result))
+    return wrap(new SwiftCachedOutputHandle(Result->getOutput(idx).Object,
+                                            Comp->DB.getCAS()));
+
+  auto *Result = std::get_if<clang::cas::CompileJobCacheResult>(&Comp->Result);
+  assert(Result && "unexpected variant");
+  return wrap(new SwiftCachedOutputHandle(Result->getOutput(idx).Object,
+                                          Comp->DB.getCAS()));
+}
+
+bool swiftscan_cached_compilation_is_uncacheable(
+    swiftscan_cached_compilation_t id) {
+  // Currently, all compilations are cacheable.
+  return false;
+}
+
+void swiftscan_cached_compilation_dispose(swiftscan_cached_compilation_t id) {
+  delete unwrap(id);
+}
+
+bool swiftscan_cached_output_load(swiftscan_cached_output_t output,
+                                  swiftscan_string_ref_t *error) {
+  auto *Out = unwrap(output);
+  // If proxy exists, there is nothing to be done.
+  if (Out->LoadedProxy)
+    return true;
+
+  auto failure = [&](llvm::Error E) {
+    *error =
+        swift::c_string_utils::create_clone(toString(std::move(E)).c_str());
+    return false;
+  };
+  if (auto Err = Out->CAS.getProxy(Out->Ref).moveInto(Out->LoadedProxy))
+    return failure(std::move(Err));
+
+  llvm::DenseSet<llvm::cas::ObjectRef> Visited;
+  llvm::SmallVector<llvm::cas::ObjectRef> WorkList;
+  auto addToWorkList = [&](llvm::cas::ObjectRef Item) {
+    if (Visited.insert(Item).second)
+      WorkList.push_back(Item);
+  };
+
+  addToWorkList(Out->LoadedProxy->getRef());
+  while (!WorkList.empty()) {
+    auto Current = WorkList.pop_back_val();
+    auto Proxy = Out->CAS.getProxyIfExists(Current);
+    if (!Proxy)
+      return failure(Proxy.takeError());
+
+    if (!*Proxy)
+      return false;
```

Comment:
Should you do
```
*error = swift::c_string_utils::create_null();
```
here ?

I've seen this in other places, should the `error` parameter always be set by the functions?